### ISSUE TITLE
Pis

### ISSUE THREAD

**Issue by stoermelder** [VCVRack/Rack#1504]
# Details

Operating system: Windows 10
Rack version: v1.1.3
Hardware relevant to your issue (e.g. graphic card model, audio/MIDI device): -

### Summary

Concurrent access to `Engine::internal->paramHandleCache` can crash Rack.

### Steps to reproduce

Repeatedly load a preset of MIDI-MAP with many mappings. 
Each mapped parameter loaded from the preset will cause `Engine::internal->paramHandleCache` to clear and rebuilt in one of the dsp-threads (with locking, no problem here).
The app-thread accesses in parallel in `ParamWidget::draw(const DrawArgs &args)` this datastructure without locking which can cause a segmentation fault when cleared at the same time.

### Possible fix

Create a copy of `Engine::internal->paramHandleCache` for the app-thread. Mark `Engine::internal->paramHandleCache` as dirty when mappings have changed and rebuilt app-thread's copy before rendering the next frame.

**Comment by AndrewBelt** [comment]
Your proposed solution will give out-of-date ParamHandles which will result in crashes.

**Comment by stoermelder** [comment]
Yes, that is correct how I described it.
But, the app-thread only needs ParamHandels to determine if the mapping indicator should be drawn and what color it should have. Considering this it would be enough to create app-thread‘s „copy“ of paramHandleCache without pointing to  instances of ParamHandle but to some new struct that only contains the needed information.
Still this information could be out-of-date but it means that mapping indicator will be drawn one additional frame.

**Comment by stoermelder** [comment]
This issue has low priority. I get many crashes because of this while debugging (compiled without optimization-flags), but I never a had crash using the release version of Rack,
